### ISSUE TITLE
not closing connections "TIME_WAIT"

### ISSUE THREAD

**Issue by Andy2244** [cifsd-team/cifsd#295]
Got this report, we send "kill_server" on stop. So should this close connections?

> smbd is also not closing connections on stop. It keep existing connection in a "TIME_WAIT" state, which prevents for a time smbd to listen again port 445.

**Comment by Andy2244** [comment]
I'm just the messenger, but i think `netstat -tulpn` and `tcptrack -i eth0` shows the status.

**Comment by namjaejeon** [comment]
Okay, When I research it, it is not problem. 
when server connection is disconnected ahead than client connection, It will be time wait state. smbd use SO_REUSEADDR, so it should re-bind on time-wait state. Is it problem ? samba is also same ?

**Comment by luizluca** [comment]
Hello, I'm the original bug reporter. Thanks @Andy2244 .

During my tests, smbd did not start while there was an existing TIME_WAIT socket. Maybe it is something specific to OpenWrt build?

**Comment by namjaejeon** [comment]
@luizluca Hello, First, Thanks for your reporting.
Could you please share how to reproduce it ?

Thanks!

**Comment by namjaejeon** [comment]
@luizluca we do recently fix SO_REUSEADDR bug.(https://github.com/cifsd-team/cifsd/commit/b2677065a00992ac882f08d98ab4d3e6b0dd5837). Could you please try to reproduce it on  the latest master of ksmbd ?

**Comment by luizluca** [comment]
> Could you please share how to reproduce it ?

Sure. I installed it on OpenWrt 19.07.1. After ksmbd 3.1.0 was up, I connected with some clients. I restarted service and got this:

```
[ 6858.873511] ksmbd: kill_server_store:437: kill command received
[ 6868.772132] ksmbd: create_socket:462: Failed to bind socket: -125
[ 6868.778495] ksmbd: tcp_destroy_socket:404: Failed to shutdown socket: -134
[ 6868.785914] Failed to init TCP subsystem: -125
[ 6883.972163] ksmbd: create_socket:462: Failed to bind socket: -125
[ 6883.978587] ksmbd: tcp_destroy_socket:404: Failed to shutdown socket: -134
[ 6883.985811] Failed to init TCP subsystem: -125
```
The fix b267706 seems to fix exactly what I'm facing. @Andy2244, do you have a recent pre-compiled package for 19.07 mips_24kc? v3.1.3 includes the fix.

**Comment by Andy2244** [comment]
> do you have a recent pre-compiled package for 19.07 mips_24kc? v3.1.3 includes the fix.

3.1.3 is "11 hours ago" i'm not a buildbot, so maybe i can make some room tomorrow, but i also have to test all the new namechanges again, so it's not a trivial PR.

**Comment by namjaejeon** [comment]
@Andy2244 please wait a bit more. I will release cifsd-tools one more tomorrow.

**Comment by Andy2244** [comment]
Sure, had no time this weekend, so next weekend is the new eta for me.

**Comment by namjaejeon** [comment]
Please pull cifsd 3.1.3 and cifsd-tools 3.2.1 version if you have time.
Thanks!

**Comment by Andy2244** [comment]
@luizluca try https://github.com/openwrt/packages/pull/11290

**Comment by luizluca** [comment]
> i'm not a buildbot

@Andy2244 , you might not be but you can configure one: luizluca/openwrt-packages@29d7601feb84551578c4a72e7b07f3c16e29a344

I tried to configure it to build nightly (I'm not a circleci expert). I'll check if it works tomorrow.
It uses 19.07.1 ath79 sdk by default (which matches what I need). You can create different branches for different stable releases, 19.07-snapshot, master-snapshot and each for different targets.

Each build will have the packages (*.ipk) as artifacts.

**Comment by Andy2244** [comment]
@luizluca yeah i have also created something via docker for a fast build: https://github.com/Andy2244/openwrt-package-builder

What i meant was that its not feasible for me to build+test every few days, so i usually reserve some time on weekends, since actually testing on live hardware is the time consuming part.

**Comment by Andy2244** [comment]
@luizluca is this still a issue with the latest builds from snapshots?

**Comment by luizluca** [comment]
I guess that one was fixed sometime ago.
Anyway, I was not able to test it as I got this error when using both kernel module
and server from current cifsd project (-next brach)

 [20838.758539] ksmbd: ksmbd_tcp_set_interfaces:596: Unknown interface:

I got that kind of error in the past when I used different kernel/userland code versions.
However, now both are freshly built. It happens even with bind interfaces off.
